synapse/storage/schema/main/delta/73/01thread_notifications_backfill.sql

```diff
@@ -0,0 +1,29 @@
+/* Copyright 2022 The Matrix.org Foundation C.I.C
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+-- Forces the background updates from 06thread_notifications.sql to run in the
+-- foreground as code will now require those to be "done".
+
+DELETE FROM background_updates WHERE update_name = 'event_push_backfill_thread_id';
+
+-- Overwrite any null thread_id columns.
+UPDATE event_push_actions_staging SET thread_id = 'main' WHERE thread_id IS NULL;
+UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
+UPDATE event_push_summary SET thread_id = 'main' WHERE thread_id IS NULL;
```

Comment:
Are these "fast" when the background update has been run?

Reply:
They should be, I'm not really sure how to check that though?

Reply:
Going to check this with a fake table to see how bad it is.

Reply:
To test this I:

1. Took a large-ish database with an `event_push_actions` table of 340,977 rows.
2. Altered the table to add the nullable `thread_id` column.
3. Ran the above twice (with `EXPLAIN ANALYZE`).

Expectations are that the first time will take some time since it is filling in all the data, while the second time it is "fast" since there's nothing to do.

<details>
<summary>Results</summary>

```sql
postgres=# EXPLAIN ANALYZE UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
                                                           QUERY PLAN                                                            
---------------------------------------------------------------------------------------------------------------------------------
 Update on event_push_actions  (cost=0.00..9712.77 rows=1705 width=229) (actual time=21128.858..21128.859 rows=0 loops=1)
   ->  Seq Scan on event_push_actions  (cost=0.00..9712.77 rows=1705 width=229) (actual time=0.099..531.200 rows=340977 loops=1)
         Filter: (thread_id IS NULL)
 Planning Time: 0.154 ms
 Execution Time: 21128.894 ms
(5 rows)

Time: 21191.277 ms (00:21.191)
postgres=# EXPLAIN ANALYZE UPDATE event_push_actions SET thread_id = 'main' WHERE thread_id IS NULL;
                                                          QUERY PLAN                                                           
-------------------------------------------------------------------------------------------------------------------------------
 Update on event_push_actions  (cost=0.00..19966.42 rows=3505 width=229) (actual time=119.550..119.551 rows=0 loops=1)
   ->  Seq Scan on event_push_actions  (cost=0.00..19966.42 rows=3505 width=229) (actual time=119.548..119.549 rows=0 loops=1)
         Filter: (thread_id IS NULL)
         Rows Removed by Filter: 340977
 Planning Time: 0.073 ms
 Execution Time: 119.598 ms
(6 rows)

Time: 122.390 ms
```

</details>

In summary, the first run through took ~20 seconds (which isn't too terrible), but the second run through was ~0.1 seconds and seems to have done some smart filtering based on `thread_id IS NULL`.